File: src/PlayerCore/WalkthroughRunner.cs
using QuestViva.Common;

namespace QuestViva.PlayerCore;

public class WalkthroughRunner(IGameDebug game, string walkthrough)
{
    public delegate Task ClearBufferEventHandler();

    public delegate Task OutputEventHandler(string text);

    private readonly IGame _game = (IGame) game;
    private bool _cancelled;
    private IDictionary<string, string>? _menuOptions;
    private bool _pausing;
    private int _queuedTickSeconds;
    private bool _showingMenu;
    private bool _showingQuestion;
    private bool _waiting;

    public int Steps => game.Walkthroughs.Walkthroughs[walkthrough].Steps.Length;

    public event OutputEventHandler? Output;

    public event ClearBufferEventHandler? ClearBuffer;

    public async Task Run()
    {
        var delay = 0;
        var dateStart = DateTime.Now;
        foreach (var cmd in game.Walkthroughs.Walkthroughs[walkthrough].Steps)
        {
            if (_cancelled)
            {
                break;
            }

            if (_showingMenu)
            {
                await SetMenuResponse(cmd);
            }
            else if (_showingQuestion)
            {
                await SetQuestionResponse(cmd);
            }
            else if (cmd.StartsWith("assert:"))
            {
                var expr = cmd.Substring(7);
                WriteLine("<br><b>Assert:</b> " + expr);
                if (await game.AssertAsync(expr))
                {
                    WriteLine("<br><span style=\"color:green\"><b>Pass</b></span>");
                }
                else
                {
                    WriteLine("<br><span style=\"color:red\"><b>Failed</b></span>");
                    return;
                }
            }
            else if (cmd.StartsWith("label:"))
            {
                // ignore
            }
            else if (cmd.StartsWith("delay:"))
            {
                delay = int.Parse(cmd[6..].Trim());
            }
            else if (cmd.StartsWith("event:"))
            {
                var eventData = cmd.Substring(6).Split([';'], 2);
                var eventName = eventData[0];
                var param = eventData[1];
                await _game.SendEvent(eventName, param);
            }
            else if (cmd.StartsWith("runtime:"))
            {
                var dateEnd = DateTime.Now;
                var diff = dateEnd.Subtract(dateStart);
                var res = $"{diff.Hours}:{diff.Minutes}:{diff.Seconds}";
                WriteLine("Walkthrough Runtime: " + res);
            }
            else
            {
                await _game.SendCommand(cmd);
            }

            await FinishStep();

            if (_cancelled)
            {
                break;
            }

            await Task.Delay(delay);
        }

        // Runs any tick that arrived during the last step, which would otherwise be dropped - and
        // the player's timer only restarts once the engine asks for the next tick.
        if (!_cancelled)
        {
            await FinishStep();
        }
    }

    // The player's real-time timer keeps running during a walkthrough, since the steps don't go
    // through its own sendCommand (which stops the timer until a turn ends). A tick arriving while
    // a step is still running would run timer scripts in the middle of it, so the player queues
    // ticks here instead, and FinishStep runs them once the step's command has finished.
    public void QueueTick(int elapsedTime)
    {
        Interlocked.Add(ref _queuedTickSeconds, elapsedTime);
    }

    // Runs queued ticks and clicks through any wait/pause the step or a timer script opened.
    private async Task FinishStep()
    {
        do
        {
            var elapsedTime = Interlocked.Exchange(ref _queuedTickSeconds, 0);
            if (elapsedTime > 0)
            {
                await game.TickUntilSuspended(elapsedTime);
            }

            if (ClearBuffer != null)
            {
                await ClearBuffer.Invoke();
            }

            if (_cancelled)
            {
                return;
            }

            if (_waiting)
            {
                _waiting = false;
                await FinishWait();
            }

            if (_pausing)
            {
                _pausing = false;
                await FinishPause();
            }
        } while ((_waiting || _pausing) && !_cancelled);
    }

    public void ShowMenu(MenuData menu)
    {
        _showingMenu = true;
        WriteLine("Menu: " + menu.Caption);
        _menuOptions = menu.Options;
    }

    private async Task SetMenuResponse(string response)
    {
        if (response.StartsWith("menu:"))
        {
            _showingMenu = false;
            var menuResponse = response[5..];
            WriteLine("  - " + menuResponse);
            if (_menuOptions!.ContainsKey(menuResponse))
            {
                await _game.SetMenuResponse(menuResponse);
            }
            else
            {
                throw new Exception("Menu response was not an option");
            }
        }
        else
        {
            throw new Exception("No menu response defined in walkthrough");
        }
    }

    public void ShowQuestion(string question, bool inline)
    {
        _showingQuestion = true;
        WriteLine("Question: " + question);
    }

    private async Task SetQuestionResponse(string response)
    {
        if (response.StartsWith("answer:"))
        {
            _showingQuestion = false;
            var questionResponse = response.Substring(7);
            WriteLine("  - " + questionResponse);
            switch (questionResponse)
            {
                case "yes":
                    await _game.SetQuestionResponse(true);
                    break;
                case "no":
                    await _game.SetQuestionResponse(false);
                    break;
                default:
                    throw new Exception("Question response was invalid");
            }
        }
        else
        {
            throw new Exception("Question response not defined in walkthrough");
        }
    }

    public void BeginWait()
    {
        _waiting = true;
    }

    private async Task FinishWait()
    {
        await _game.FinishWait();
    }

    public void BeginPause()
    {
        _pausing = true;
    }

    private async Task FinishPause()
    {
        await _game.FinishPause();
    }

    private void WriteLine(string text)
    {
        Output?.Invoke(text);
    }

    public void Cancel()
    {
        _cancelled = true;
    }
}

File: legacy/Player/WalkthroughRunner.vb
﻿Imports System.Threading

Friend Class WalkthroughRunner
    Private m_gameDebug As IASLDebug
    Private m_game As IASL
    Private m_walkthrough As String
    Private m_showingMenu As Boolean
    Private m_showingQuestion As Boolean
    Private m_waiting As Boolean
    Private m_pausing As Boolean
    Private m_menuOptions As IDictionary(Of String, String)
    Private m_cancelled As Boolean

    Public Event Output(text As String)
    Public Event MarkScrollPosition()

    Public Sub New(game As IASLDebug, walkthrough As String)
        m_gameDebug = game
        m_game = DirectCast(game, IASL)
        m_walkthrough = walkthrough
    End Sub

    Public Sub Run()
        Dim delay As Integer = 0
        Dim dateStart = DateTime.Now
        For Each cmd As String In m_gameDebug.Walkthroughs.Walkthroughs(m_walkthrough).Steps
            If m_cancelled Then Exit For
            RaiseEvent MarkScrollPosition()
            If m_showingMenu Then
                SetMenuResponse(cmd)
            ElseIf m_showingQuestion Then
                SetQuestionResponse(cmd)
            Else
                If cmd.StartsWith("assert:") Then
                    Dim expr As String = cmd.Substring(7)
                    WriteLine("<br/><b>Assert:</b> " + expr)
                    If m_gameDebug.Assert(expr) Then
                        WriteLine("<span style=""color:green""><b>Pass</b></span>")
                    Else
                        WriteLine("<span style=""color:red""><b>Failed</b></span>")
                        Return
                    End If
                ElseIf cmd.StartsWith("label:") Then
                    ' ignore
                ElseIf cmd.StartsWith("delay:") Then
                    delay = Int32.Parse(cmd.Substring(6).Trim)
                ElseIf cmd.StartsWith("event:") Then
                    Dim eventData As String() = cmd.Substring(6).Split(New Char() {";"c}, 2)
                    Dim eventName As String = eventData(0)
                    Dim param As String = eventData(1)
                    m_game.SendEvent(eventName, param)
                ElseIf cmd.StartsWith("runtime:") Then
                    Dim dateEnd = DateTime.Now
                    Dim diff = dateEnd.Subtract(dateStart)
                    Dim res = String.Format("{0}:{1}:{2}", diff.Hours, diff.Minutes, diff.Seconds)
                    WriteLine("Walkthrough Runtime: " + res)
                Else
                    m_game.SendCommand(cmd)
                End If
            End If

            If m_cancelled Then Exit For

            Do
                If m_waiting Then
                    m_waiting = False
                    FinishWait()
                End If

                If m_pausing Then
                    m_pausing = False
                    FinishPause()
                End If
            Loop Until (Not m_waiting And Not m_pausing) Or m_cancelled
            Thread.Sleep(delay)
        Next

    End Sub

    Public Sub ShowMenu(menu As MenuData)
        m_showingMenu = True
        WriteLine("Menu: " + menu.Caption)
        m_menuOptions = menu.Options
    End Sub

    Private Sub SetMenuResponse(response As String)
        If response.StartsWith("menu:") Then
            m_showingMenu = False
            Dim menuResponse As String = response.Substring(5)
            WriteLine("  - " + menuResponse)
            If (m_menuOptions.ContainsKey(menuResponse)) Then
                m_game.SetMenuResponse(menuResponse)
            Else
                Throw New Exception("Menu response was not an option")
            End If
        Else
            Throw New Exception("No menu response defined in walkthrough")
        End If
    End Sub

    Public Sub ShowQuestion(question As String)
        m_showingQuestion = True
        WriteLine("Question: " + question)
    End Sub

    Private Sub SetQuestionResponse(response As String)
        If response.StartsWith("answer:") Then
            m_showingQuestion = False
            Dim questionResponse As String = response.Substring(7)
            WriteLine("  - " + questionResponse)
            If questionResponse = "yes" Then
                m_game.SetQuestionResponse(True)
            ElseIf questionResponse = "no" Then
                m_game.SetQuestionResponse(False)
            Else
                Throw New Exception("Question response was invalid")
            End If
        Else
            Throw New Exception("Question response not defined in walkthrough")
        End If
    End Sub

    Public Sub BeginWait()
        m_waiting = True
    End Sub

    Private Sub FinishWait()
        m_game.FinishWait()
    End Sub

    Public Sub BeginPause()
        m_pausing = True
    End Sub

    Private Sub FinishPause()
        m_game.FinishPause()
    End Sub

    Private Sub WriteLine(text As String)
        RaiseEvent Output(text)
    End Sub

    Public ReadOnly Property Steps As Integer
        Get
            Return m_gameDebug.Walkthroughs.Walkthroughs(m_walkthrough).Steps.Count
        End Get
    End Property

    Public Sub Cancel()
        m_cancelled = True
    End Sub
End Class
